File: Assets/Scripts/Pose/PoseData.cs
using UnityEngine;

namespace FitDungeon.Pose
{
    /// <summary>
    /// 姿态数据 - 存储人体关键点信息
    /// </summary>
    public class PoseData
    {
        // MediaPipe 33个关键点
        public Vector2[] Keypoints { get; private set; }
        public float[] Confidences { get; private set; }
        public bool IsValid { get; private set; }
        
        // 关键点索引（MediaPipe标准）
        public const int NOSE = 0;
        public const int LEFT_SHOULDER = 11;
        public const int RIGHT_SHOULDER = 12;
        public const int LEFT_ELBOW = 13;
        public const int RIGHT_ELBOW = 14;
        public const int LEFT_WRIST = 15;
        public const int RIGHT_WRIST = 16;
        public const int LEFT_HIP = 23;
        public const int RIGHT_HIP = 24;
        public const int LEFT_KNEE = 25;
        public const int RIGHT_KNEE = 26;
        public const int LEFT_ANKLE = 27;
        public const int RIGHT_ANKLE = 28;
        
        public PoseData()
        {
            Keypoints = new Vector2[33];
            Confidences = new float[33];
            IsValid = false;
        }
        
        /// <summary>
        /// 更新关键点数据
        /// </summary>
        public void UpdateKeypoint(int index, Vector2 position, float confidence)
        {
            if (index >= 0 && index < 33)
            {
                Keypoints[index] = position;
                Confidences[index] = confidence;
            }
        }
        
        /// <summary>
        /// 获取关键点（带置信度检查）
        /// </summary>
        public bool TryGetKeypoint(int index, out Vector2 position, float minConfidence = 0.5f)
        {
            position = Vector2.zero;
            if (index >= 0 && index < 33 && Confidences[index] >= minConfidence)
            {
                position = Keypoints[index];
                return true;
            }
            return false;
        }
        
        /// <summary>
        /// 计算两点之间的角度
        /// </summary>
        public float CalculateAngle(int point1, int point2, int point3)
        {
            Vector2 p1, p2, p3;
            
            if (!TryGetKeypoint(point1, out p1) ||
                !TryGetKeypoint(point2, out p2) ||
                !TryGetKeypoint(point3, out p3))
            {
                return -1f; // 无效
            }
            
            Vector2 v1 = p1 - p2;
            Vector2 v2 = p3 - p2;
            
            float angle = Vector2.Angle(v1, v2);
            return angle;
        }
        
        /// <summary>
        /// 标记数据为有效/无效
        /// </summary>
        public void SetValid(bool valid)
        {
            IsValid = valid;
        }
    }
}

File: lib/models/pose_data.dart
import 'dart:ui';
import 'package:google_mlkit_pose_detection/google_mlkit_pose_detection.dart';

/// 姿态数据模型
/// 封装 ML Kit 的姿态检测结果
class PoseData {
  /// 所有关键点
  final Map<PoseLandmarkType, PoseLandmarkData> landmarks;
  
  /// 图像尺寸（用于归一化）
  final Size imageSize;
  
  PoseData({
    required this.landmarks,
    required this.imageSize,
  });
  
  /// 从 ML Kit Pose 创建
  factory PoseData.fromPose(Pose pose, [Size imageSize = const Size(1, 1)]) {
    final landmarks = <PoseLandmarkType, PoseLandmarkData>{};
    
    // 映射所有关键点
    for (final type in PoseLandmarkType.values) {
      final landmark = pose.landmarks[type];
      if (landmark != null) {
        landmarks[type] = PoseLandmarkData(
          x: landmark.x,
          y: landmark.y,
          z: landmark.z,
          likelihood: landmark.likelihood,
        );
      }
    }
    
    return PoseData(landmarks: landmarks, imageSize: imageSize);
  }
  
  /// 获取指定关键点
  PoseLandmarkData? getLandmark(PoseLandmarkType type) => landmarks[type];
  
  /// 获取归一化的关键点坐标 (0-1)
  Offset? getNormalizedPosition(PoseLandmarkType type) {
    final landmark = landmarks[type];
    if (landmark == null || imageSize == Size.zero) return null;
    
    return Offset(
      landmark.x / imageSize.width,
      landmark.y / imageSize.height,
    );
  }
  
  /// 检查关键点是否有效
  bool hasLandmark(PoseLandmarkType type) {
    final landmark = landmarks[type];
    return landmark != null && landmark.likelihood > 0.5;
  }
  
  // 常用关键点便捷访问器
  
  PoseLandmarkData? get nose => getLandmark(PoseLandmarkType.nose);
  PoseLandmarkData? get leftShoulder => getLandmark(PoseLandmarkType.leftShoulder);
  PoseLandmarkData? get rightShoulder => getLandmark(PoseLandmarkType.rightShoulder);
  PoseLandmarkData? get leftElbow => getLandmark(PoseLandmarkType.leftElbow);
  PoseLandmarkData? get rightElbow => getLandmark(PoseLandmarkType.rightElbow);
  PoseLandmarkData? get leftWrist => getLandmark(PoseLandmarkType.leftWrist);
  PoseLandmarkData? get rightWrist => getLandmark(PoseLandmarkType.rightWrist);
  PoseLandmarkData? get leftHip => getLandmark(PoseLandmarkType.leftHip);
  PoseLandmarkData? get rightHip => getLandmark(PoseLandmarkType.rightHip);
  PoseLandmarkData? get leftKnee => getLandmark(PoseLandmarkType.leftKnee);
  PoseLandmarkData? get rightKnee => getLandmark(PoseLandmarkType.rightKnee);
  PoseLandmarkData? get leftAnkle => getLandmark(PoseLandmarkType.leftAnkle);
  PoseLandmarkData? get rightAnkle => getLandmark(PoseLandmarkType.rightAnkle);
  
  /// 计算两个关键点之间的角度
  double? getAngle(
    PoseLandmarkType first,
    PoseLandmarkType mid,
    PoseLandmarkType last,
  ) {
    final firstPoint = landmarks[first];
    final midPoint = landmarks[mid];
    final lastPoint = landmarks[last];
    
    if (firstPoint == null || midPoint == null || lastPoint == null) {
      return null;
    }
    
    // 计算向量
    final vector1 = Offset(
      firstPoint.x - midPoint.x,
      firstPoint.y - midPoint.y,
    );
    final vector2 = Offset(
      lastPoint.x - midPoint.x,
      lastPoint.y - midPoint.y,
    );
    
    // 计算角度
    final dot = vector1.dx * vector2.dx + vector1.dy * vector2.dy;
    final mag1 = vector1.distance;
    final mag2 = vector2.distance;
    
    if (mag1 == 0 || mag2 == 0) return null;
    
    final cos = (dot / (mag1 * mag2)).clamp(-1.0, 1.0);
    return (180 / 3.14159) * acos(cos);
  }
}

/// 关键点数据
class PoseLandmarkData {
  final double x;
  final double y;
  final double z;
  final double likelihood;
  
  PoseLandmarkData({
    required this.x,
    required this.y,
    required this.z,
    required this.likelihood,
  });
  
  Offset get position => Offset(x, y);
}
